NuKeeper.Gitlab/GitlabRestClient.cs

```diff
@@ -0,0 +1,137 @@
+            string branchName)
+        {
+            var encodedProjectName = HttpUtility.UrlEncode($"{projectName}/{repositoryName}");
+            var encodedBranchName = HttpUtility.UrlEncode(branchName);
+
+            return await GetResource<Model.Branch>(
+                $"projects/{encodedProjectName}/repository/branches/{encodedBranchName}",
+                statusCode => statusCode == HttpStatusCode.NotFound
+                    ? Result<Model.Branch>.Success(null)
+                    : Result<Model.Branch>.Failure()).ConfigureAwait(false);
+        }
+
+        // https://docs.gitlab.com/ee/api/merge_requests.html#create-mr
+        // POST /projects/:id/merge_requests
+        public Task<Model.MergeRequest> OpenMergeRequest(string projectName, string repositoryName, Model.MergeRequest mergeRequest)
+        {
+            var encodedProjectName = HttpUtility.UrlEncode($"{projectName}/{repositoryName}");
+
+            var content = new StringContent(JsonConvert.SerializeObject(mergeRequest), Encoding.UTF8,
+                "application/json");
+            return PostResource<Model.MergeRequest>($"projects/{encodedProjectName}/merge_requests", content);
+        }
+
+        private async Task<T> GetResource<T>(string url, Func<HttpStatusCode, Result<T>> customErrorHandling = null, [CallerMemberName] string caller = null)
+        {
+            var fullUrl = new Uri(url, UriKind.Relative);
+            _logger.Detailed($"{caller}: Requesting {fullUrl}");
+
+            var response = await _client.GetAsync(fullUrl).ConfigureAwait(false);
+            return await HandleResponse<T>(response, customErrorHandling, caller).ConfigureAwait(false);
+        }
+
+        private async Task<T> PostResource<T>(string url, HttpContent content, Func<HttpStatusCode, Result<T>> customErrorHandling = null, [CallerMemberName] string caller = null)
+        {
+            _logger.Detailed($"{caller}: Requesting {url}");
+
+            var response = await _client.PostAsync(url, content).ConfigureAwait(false);
+
+            return await HandleResponse<T>(response, customErrorHandling, caller).ConfigureAwait(false);
+        }
+
+        private async Task<T> HandleResponse<T>(HttpResponseMessage response,
+            Func<HttpStatusCode, Result<T>> customErrorHandling,
+            [CallerMemberName] string caller = null)
+        {
+            string msg;
+
+            var responseBody = await response.Content.ReadAsStringAsync().ConfigureAwait(false);
+
+            if (!response.IsSuccessStatusCode)
+            {
+                _logger.Detailed($"Response {response.StatusCode} is not success, body:\n{responseBody}");
+
+                if (customErrorHandling != null)
+                {
+                    var result = customErrorHandling(response.StatusCode);
+                    if (result.IsSuccessful)
+                        return result.Value;
```

Comment:
I love F# and `Result` type, but I'm finding the usage here a little odd. Seemingly this is the only place where it's unpacked, there's only one call to here where `customErrorHandling` is not `null`, the name `customErrorHandling` seems off as it's mapping a status code to a result, and the `Result` type should instead have a `Match` method take takes success and failure `Func`s IMO.

Reply:
Interesting. I have used result wrapper types many times, idea taken from too many places to mention.  It's definitely applicable outside of just GitLab integration, and can have more functional extensions, but that can come in later refactoring.

Reply:
What would be your suggestion here, guys?
Currently it's here just to handle 404 coming from the API. That's a valid response for non existing branch.

I can probably move the `Result` type to a `Nukeeper.Abstractons` assembly and make it more consistent with i.e. F# implementation.

Reply:
On further reflection, what is there currently is fine, it's a private class so it can always be changed.

It doesn't have to look like the F# implementation, but I found myself unfamiliar with what I expected (I'm easily confused). However, looking at the one place it's called it makes no sense to add a `.Match(Action<T> onSuccess, Action onFailure)` or `.WithDefault(T defaultValue)`, because we want to branch and conditionally return anyway.

My only last thought is that it might be better suited to a tuple return `(bool isSuccessful, T result)`, but I'm just nit-picking 🙂

Reply:
Thinking it might be a bit confusing to get a tuple containing both `(bool isSuccessful, T result)`, since result only makes sense when `isSuccessful` is true.
With current implementation trying to get the `Value` will throw an exception at you when `IsSuccessful` is false. 🤔 

Let me know what you think. 🙂

Reply:
It's the problem as the Result type here, except it throws a more explicit exception. Ignore me, I'm not progressing anything here 🙂